src/terminal/default.ts

```diff
@@ -0,0 +1,358 @@
+        resolve(this);
+      };
+      this._ws.onerror = (event: Event) => {
+        reject(this);
+      };
+    });
+  }
+
+  private _name: string;
+  private _baseUrl: string;
+  private _wsUrl: string;
+  private _url: string;
+  private _ajaxSettings = '';
+  private _ws: WebSocket = null;
+  private _isDisposed = false;
+  private _promise: Promise<TerminalSession.ISession>;
+}
+
+
+
+export
+namespace DefaultTerminalSession {
+  /**
+   * Start a new terminal session.
+   *
+   * @options - The session options to use.
+   *
+   * @returns A promise that resolves with the session instance.
+   */
+  export
+  function startNew(options: TerminalSession.IOptions = {}): Promise<TerminalSession.ISession> {
+    return new DefaultTerminalSession('', options).connect();
+  }
+
+  /*
+   * Connect to a running session.
+   *
+   * @param name - The name of the target session.
+   *
+   * @param options - The session options to use.
+   *
+   * @returns A promise that resolves with the new session instance.
+   *
+   * #### Notes
+   * If the session was already started via `startNew`, the existing
+   * session object is used as the fulfillment value.
+   *
+   * Otherwise, if `options` are given, we attempt to connect to the existing
+   * session.
+   * The promise is fulfilled when the session is ready on the server,
+   * otherwise the promise is rejected.
+   *
+   * If the session was not already started and no `options` are given,
+   * the promise is rejected.
+   */
+  export
+  function connectTo(name: string, options: TerminalSession.IOptions = {}): Promise<TerminalSession.ISession> {
+    options.baseUrl = options.baseUrl || utils.getBaseUrl();
+    let url = Private.getTermUrl(options.baseUrl, name);
+    if (url in Private.running) {
+      return Private.running[url].connect();
+    }
+    return new DefaultTerminalSession(name, options).connect();
+  }
+
+  /**
+   * List the running terminal sessions.
+   *
+   * @param options - The session options to use.
+   *
+   * @returns A promise that resolves with the list of running session models.
+   */
+  export
+  function listRunning(options: TerminalSession.IOptions = {}): Promise<TerminalSession.ISession[]> {
+    let url = Private.getBaseUrl(options.baseUrl);
+    let ajaxSettings = utils.copy(options.ajaxSettings || {}) as IAjaxSettings;
+    ajaxSettings.method = 'GET';
+    ajaxSettings.dataType = 'json';
+
+    return utils.ajaxRequest(url, ajaxSettings).then(success => {
+      if (success.xhr.status !== 200) {
+        return utils.makeAjaxError(success);
+      }
+      let data = success.data as TerminalSession.IModel[];
+      if (!Array.isArray(data)) {
+        return utils.makeAjaxError(success, 'Invalid terminal data');
+      }
+      // Update the local data store.
+      let urls = toArray(map(data, item => {
+          return utils.urlPathJoin(url, item.name);
+      }));
+      each(Object.keys(Private.running), runningUrl => {
```

Comment:
*e.g.* if `Private.running` were a `Set` you could do `Private.running.forEach(...)`